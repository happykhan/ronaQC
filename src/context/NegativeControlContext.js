import React from "react";

const NegativeControlContext = React.createContext();

export default NegativeControlContext;
