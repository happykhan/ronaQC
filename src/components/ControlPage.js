import React, { useContext } from "react";
import {
  Typography,
  Container,
  Card,
  CardContent,
  Table,
  CircularProgress,
  TableBody,
  TableHead,
  TableCell,
  TableRow,
} from "@mui/material/";
import NegativeControlContext from "../context/NegativeControlContext";
import Covplot from "./CovPlot";
import ErrorIcon from "@mui/icons-material/Error";
import WarningIcon from "@mui/icons-material/Warning";
import CheckCircleIcon from "@mui/icons-material/CheckCircle";
const SummaryTable = (control) => {
  return (
    <Table sx={{ minWidth: 400 }} aria-label="NC table">
      <TableHead>
        <TableRow>
          <TableCell>
            SNPs {control.snpCount > 0 && <ErrorIcon sx={{ color: "red" }} />}
            {control.snpCount === 0 && (
              <CheckCircleIcon sx={{ color: "green" }} />
            )}
          </TableCell>
          <TableCell>
            Called bases
            {control.genomeRecovery >= 500 && (
              <WarningIcon sx={{ color: "orange" }} />
            )}
            {control.genomeRecovery < 500 && (
              <CheckCircleIcon sx={{ color: "green" }} />
            )}
          </TableCell>
          <TableCell>Best mapped reads / Well mapped reads</TableCell>
          <TableCell>
            Detected amplicons{" "}
            {(control.amplicons ? control.amplicons.length : 0) > 1 && (
              <ErrorIcon sx={{ color: "red" }} />
            )}{" "}
            {(control.amplicons ? control.amplicons.length : 0) == 1 && (
              <WarningIcon sx={{ color: "orange" }} />
            )}
            {(control.amplicons ? control.amplicons.length : 0) == 0 && (
              <CheckCircleIcon sx={{ color: "green" }} />
            )}
          </TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        <TableRow>
          <TableCell>
            {control.snpCount ? control.snpCount : "Unknown SNPs"}
          </TableCell>
          <TableCell>
            {control.genomeRecovery
              ? `${control.genomeRecovery}`
              : "Unknown genome recovery"}
          </TableCell>
          <TableCell>
            {control.onefoureight
              ? control.onefoureight
              : "Unknown mapped reads"}{" "}
            /{" "}
            {control.properReads ? control.properReads : "Unknown mapped reads"}
          </TableCell>
          <TableCell>
            {control.amplicons
              ? control.amplicons.length > 0
                ? control.amplicons.join(", ")
                : "No amplicons found"
              : "Unknown amplicons"}
          </TableCell>
        </TableRow>
      </TableBody>
    </Table>
  );
};

const ControlPage = () => {
  const { negativeControl } = useContext(NegativeControlContext);
  return (
    <Container maxWidth="md" data-testid={`Control-page`}>
      <Card>
        <CardContent>
          <Typography variant="h3" my={3} gutterBottom>
            Control report
          </Typography>
          {negativeControl.name ? (
            <Typography variant="body1">
              {`Loaded file ${negativeControl.name}`}
            </Typography>
          ) : (
            <Typography variant="body1">
              You have no control file loaded. Please add a file under Import
            </Typography>
          )}
          {negativeControl.name && <SummaryTable {...negativeControl} />}
        </CardContent>
        <CardContent>
          {negativeControl.coverage ? (
            <Covplot coverage={negativeControl.coverage} />
          ) : (
            negativeControl.name && <CircularProgress />
          )}
        </CardContent>
      </Card>
    </Container>
  );
};
export default ControlPage;
