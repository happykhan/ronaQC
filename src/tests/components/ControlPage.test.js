/**
 * @jest-environment jsdom
 */
import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import NegativeControlContext from "../../context/NegativeControlContext";
import ControlPage from "../../components/ControlPage";

const customRender = (ui, { providerProps, ...renderOptions }) => {
  return render(
    <NegativeControlContext.Provider {...providerProps}>
      {ui}
    </NegativeControlContext.Provider>,
    renderOptions
  );
};

test("should render Control page", () => {
  const negativeControl = {
    name: "test.bam",
    genomeRecovery: 400,
    snpCount: 1,
  };
  const dispatch = jest.fn();

  const providerProps = {
    value: {
      negativeControl,
      dispatch,
    },
  };

  customRender(<ControlPage />, { providerProps });
  screen.debug();
  expect(screen.getByTestId(`Control-page`)).toMatchSnapshot();
});
