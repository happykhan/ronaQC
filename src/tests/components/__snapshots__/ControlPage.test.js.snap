// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Control page 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthMd css-xdmu94-MuiContainer-root"
  data-testid="Control-page"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom css-pxxorf-MuiTypography-root"
      >
        Control report 
      </h3>
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Loaded file test.bam
      </p>
      <table
        aria-label="NC table"
        class="MuiTable-root css-vmwj21-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              SNPs 
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1idedao-MuiSvgIcon-root"
                data-testid="ErrorIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
                />
              </svg>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Called bases
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15yr02a-MuiSvgIcon-root"
                data-testid="CheckCircleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Best mapped reads / Well mapped reads / Total reads
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Detected amplicons
               
               
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15yr02a-MuiSvgIcon-root"
                data-testid="CheckCircleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              400
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Unknown mapped reads
               
              /
               
              Unknown mapped reads
               
              / 
              Unknown total reads
               
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Unknown amplicons
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
